File: clojure/src/ch1/ex121.clj
;; (load-file "filename.clj")
;; or on command line> clj filename.clj


(defn divides? [a b]
  (= (rem b a) 0))


(defn square [n]
  (* n n))


(defn find-divisor [n test-divisor]
  (cond (> (square test-divisor) n) n
        (divides? test-divisor n) test-divisor
        :else (find-divisor n (inc test-divisor))))


(defn smallest-divisor [n]
  (find-divisor n 2))


(require
 'clojure.test)

(clojure.test/is (= (smallest-divisor 199) 199))
(clojure.test/is (= (smallest-divisor 1999) 1999))
(clojure.test/is (= (smallest-divisor 19999) 7))

(println "done")

File: octave/ch1/Ex121.m
function Ex121
  printf("%d\n", smallest_divisor(199));
  printf("%d\n", smallest_divisor(1999));
  printf("%d\n", smallest_divisor(19999));
  printf("done\n");
endfunction


function VAL = find_divisor(n, test)
  divides = @(a, b) rem(b, a) == 0;
  
  if (test * test > n)
    VAL = n;
  elseif (divides(test, n))
    VAL = test;
  else
    VAL = find_divisor(n, test + 1);
  endif
endfunction

function VAL = smallest_divisor(n) 
  VAL = find_divisor(n, 2);
endfunction
